File: Simulation_MonteHall.py
import random
import numpy as Np
random.seed(1)

Rounds = 10000

Wins_with_no_switch = [0]*Rounds
Wins_with_switch = [0]*Rounds

# Simulating Process
for Process in range(0,Rounds):

	Doors = [i for i in range(1,4)]
	Door_with_Prize = random.randint(1,3)
	Door_chosen_by_user = random.randint(1,3)	


	# Now choose a door which is still close
	Door_chosen_by_host = random.randint(1,3)	
	while(Door_chosen_by_host==Door_chosen_by_user or Door_chosen_by_host==Door_with_Prize):
		Door_chosen_by_host = random.randint(1,3)


	# if previously the door which we choose contains prize
	if Door_chosen_by_user==Door_with_Prize:
		Wins_with_no_switch[Process] = 1 


	# switching door
	# as suppose 1, 3 are chosen already then 6 - (1+3) = 2 is remaining door
	Door_chosen_after_switching = 6 - Door_chosen_by_host - Door_chosen_by_user


	if Door_chosen_after_switching==Door_with_Prize:
		Wins_with_switch[Process] = 1


# print("Wins_with_no_switch : ", Wins_with_no_switch)
# print("Wins_with_switch : ", Wins_with_switch)
# print()

print("Probability : [Wins_with_no_switch] : [ ", Np.mean(Wins_with_no_switch)," ]")
print("Probability : [Wins_with_switch] : [ ", Np.mean(Wins_with_switch)," ]")

File: Simulation_Monte_hall.cpp
#include <bits/stdc++.h>
using namespace std;

string sp=" ";
#define all(x) x.begin(),x.end()
#define Time cout <<"\nTime : "<<1000*clock()/CLOCKS_PER_SEC<<"ms\n";

int Random(int a, int b) {
    return a + rand() % (b - a + 1);
}

int Rounds = 50; 

vector<int> Wins_with_no_switch(Rounds+1, 0);
vector<int> Wins_with_switching(Rounds+1, 0);

int Wins_no_switch = 0;
int Wins_by_switch = 0;

void Simulate(){
    for(int Process = 1; Process<=Rounds; Process += 1){

        // Basically the gameshow has 3 doors
        set<int> Doors = {1,2,3};

        // Door with the prize will be chosen randomly
        int Door_with_Prize = Random(1,3);

        // Door, which the user Chooses will also be chosen randomly
        int Door_chosen = Random(1,3);

        // Now, we have to choose a door which is closed
        // Host will not open a door which has Prize in it
        int Door_Chosen_by_Host = Random(1,3);

        // choosing the door with no prize and different from the user.
        while(Door_Chosen_by_Host==Door_chosen || Door_Chosen_by_Host==Door_with_Prize){
            Door_Chosen_by_Host = Random(1,3);
        }

        // Now we will check if the user wins or not
        if( Door_chosen == Door_with_Prize ){
            Wins_with_no_switch[Process] += 1;
            Wins_no_switch += 1;
        }

        // If the user chooses to switch then 
        Door_chosen = 6 - (Door_chosen + Door_Chosen_by_Host);

        if( Door_chosen == Door_with_Prize ){
            Wins_by_switch += 1;
            Wins_with_switching[Process] += 1;
        }
    }

    //  Statistics

    long double Prob_with_no_switching = (Wins_no_switch*1.0)/Rounds;
    long double Prob_with____switching = (Wins_by_switch*1.0)/Rounds;

    cout << "Prob_with____switching = " << Prob_with____switching << endl;
    cout << "Prob_with_no_switching = " << Prob_with_no_switching << endl;
}

int32_t main(){
    srand(01);
    cout<<fixed<<setprecision(25);
    Simulate();
    Time;
    return 0;
}
